python/hail/methods/linear_mixed_model.py

```diff
@@ -0,0 +1,675 @@
+        self._xdx_alt = np.zeros((f + 1, f + 1))
+
+        self._residual_sq = None
+
+        self._scala_model = None
+
+    def _reset(self):
+        self._fitted = False
+
+        self.beta = None
+        self.sigma_sq = None
+        self.tau_sq = None
+        self.gamma = None
+        self.log_gamma = None
+        self.h_sq = None
+        self.optimize_result = None
+
+    def compute_neg_log_reml(self, log_gamma, return_parameters=False):
+        r"""Compute negative log REML constrained to a fixed value
+        of :math:`\log{\gamma}`.
+
+        This function computes the triple :math:`(\beta, \sigma^2, \tau^2)` with
+        :math:`\gamma = \frac{\sigma^2}{\tau^2}` at which the restricted likelihood
+        is maximized and returns the negative of the log of the
+        restricted likelihood at these parameters, shifted by a constant whose value is
+        independent of the input.
+
+        To compute the actual negative log REML, add
+
+        .. math::
+
+            \frac{1}{2}\left((n - p)(1 + \log(2\pi)) - \log(\det(X^T X)\right)
+
+        to the returned value.
+
+        Parameters
+        ----------
+        log_gamma: :obj:`float`
+            Value of :math:`\log{\gamma}`.
+        return_parameters:
+            If ``True``, also return :math:`\beta`, :math:`\sigma^2`,
+            and :math:`\tau^2`.
+
+        Returns
+        -------
+        :obj:`float` or (:obj:`float`, :class:`numpy.ndarray`, :obj:`float`, :obj:`float`)
+            If `return_parameters` is ``False``, returns (shifted) negative log REML.
+            Otherwise, returns (shifted) negative log REML, :math:`\beta`, :math:`\sigma^2`,
+            and :math:`\tau^2`.
+        """
+        from scipy.linalg import solve, LinAlgError
+
+        gamma = np.exp(log_gamma)
+        d = 1 / (self.s + 1 / gamma)
+        logdet_d = np.sum(np.log(d)) + (self.n - self.r) * log_gamma
+
+        if self.low_rank:
+            d -= gamma
+            dpy = d * self.py
+            ydy = self.py @ dpy + gamma * self._yty
+            xdy = self.px.T @ dpy + gamma * self._xty
+            xdx = (self.px.T * d) @ self.px + gamma * self._xtx
+        else:
+            dpy = d * self.py
+            ydy = self.py @ dpy
+            xdy = self.px.T @ dpy
+            xdx = (self.px.T * d) @ self.px
+
+        try:
+            beta = solve(xdx, xdy, assume_a='pos')
+            residual_sq = ydy - xdy.T @ beta
+            sigma_sq = residual_sq / self._dof
+            tau_sq = sigma_sq / gamma
+            neg_log_reml = (np.linalg.slogdet(xdx)[1] - logdet_d + self._dof * np.log(sigma_sq)) / 2
+
+            if return_parameters:
+                return neg_log_reml, beta, sigma_sq, tau_sq
+            else:
+                self._d, self._ydy, self._xdy, self._xdx = d, ydy, xdy, xdx
+                return neg_log_reml
+        except LinAlgError as e:
+            raise Exception(f'linear algebra error while solving for REML estimate:\n  {e}')
```

Comment:
`raise Exception('linear algebra error while solving for REML estimate') from e`

Reply:
raise ... from ... keeps the stack trace intact, which is great.